### DISCUSSION TITLE
hcxdumptool on a notebook

### DISCUSSION THREAD

**Discussion by v74863** [ZerBea/hcxdumptool#470] [General]
Hi, Mike
Please review how I use hcxdumptool on a notebook.
Maybe you can advice something usefull.
Thank you in advance.

My main OS is still Windows, I usually run Arch Linux from an iso-file using the GRUB loader.
Initially, I used the monthly official Arch iso.
https://hashcat.net/forum/thread-11349.html
Now I am creating my own iso using the archiso tool.
My iso is the same monthly official iso-file
plus additional software like lts-kernel, hcxdumptool, etc.

I run hcxdumptool in alacritty (a terminal emulator) under i3 (a tiling window manager).
Before running hcxdumptool I disable unnecessary (in my opinion) services.
BTW: Instead of `reflector.service` I manually enter my favorite Arch mirror to `/etc/pacman.d/mirrorlist `, instead of `systemd-resolved.service` I use simple text file `/etc/resolv.conf`
```
sudo systemctl disable --now reflector.service
sudo systemctl disable --now ModemManager.service
sudo systemctl disable --now systemd-resolved.service
sudo systemctl stop systemd-networkd.socket && sudo systemctl disable --now systemd-networkd.service
sudo systemctl disable --now iwd.service
```
So I have 14 services
```
% sudo systemctl --type=service --state=running
  UNIT                        LOAD   ACTIVE SUB     DESCRIPTION
  dbus-broker.service         loaded active running D-Bus System Message Bus
  getty@tty1.service          loaded active running Getty on tty1
  getty@tty2.service          loaded active running Getty on tty2
  polkit.service              loaded active running Authorization Manager
  sshd.service                loaded active running OpenSSH Daemon
  systemd-homed.service       loaded active running Home Area Manager
  systemd-journald.service    loaded active running Journal Service
  systemd-logind.service      loaded active running User Login Management
  systemd-nsresourced.service loaded active running Namespace Resource Manager
  systemd-timesyncd.service   loaded active running Network Time Synchronization
  systemd-udevd.service       loaded active running Rule-based Manager for Device Events and Files
  systemd-userdbd.service     loaded active running User Database Manager
  user@0.service              loaded active running User Manager for UID 0
  user@1000.service           loaded active running User Manager for UID 1000

Legend: LOAD   → Reflects whether the unit definition was properly loaded.
        ACTIVE → The high-level unit activation state, i.e. generalization of SUB.
        SUB    → The low-level unit activation state, values depend on unit type.

14 loaded units listed.
```
However, after that hcxdumptool cannot find interface
```
% hcxdumptool -L

Requesting physical interface capabilities. This may take some time.
Please be patient...

failed to get interface list
```
After the command
`sudo iw phy phy0 interface add wlan0 type station`

The interface was found
```
% hcxdumptool -L

Requesting physical interface capabilities. This may take some time.
Please be patient...

available wlan devices:

phy idx hw-mac       virtual-mac  m ifname           driver (protocol)
---------------------------------------------------------------------------------------------
  0   9 112233445566 112233445566 * wlan0            ath9k (NETLINK)

* active monitor mode available (reported by driver - do not trust it)
+ monitor mode available (reported by driver)
- no monitor mode available
```
BTW: hw-mac and virtual-mac are the same so far.

Now I can scan and attack
```
sudo hcxdumptool -i wlan0 --rcascan=a --rds=1 --tot=2 -c 1a,2a,3a,4a,5a,6a,7a,8a,9a,10a,11a,12a,13a

% hcxdumptool -L

Requesting physical interface capabilities. This may take some time.
Please be patient...

available wlan devices:

phy idx hw-mac       virtual-mac  m ifname           driver (protocol)
---------------------------------------------------------------------------------------------
  0  11 112233445566 8c8401d9b055 * wlan0            ath9k (NETLINK)

* active monitor mode available (reported by driver - do not trust it)
+ monitor mode available (reported by driver)
- no monitor mode available


sudo hcxdumptool -i wlan0 -w dump.pcapng -c 5a,6a
```
If I need access to Internet, I type in separate tile of alacritty
`sudo /lib/iwd/iwd`
Then connect
```
iwctl station wlan0 scan
iwctl station wlan0 get-networks
iwctl --passphrase 12345678 station wlan0 connect AP
```
If I need hcxdumptool again,
1st, I need to cancel `iwd` (I enter Ctrl-C in alacritty tile with iwd)
2nd, I again enter
`sudo iw phy phy0 interface add wlan0 type station`

and then hcxdumptool should work as intended.
All of the above assumes my current notebook with a single built-in wifi card.

**Comment by ZerBea** [comment]
there is no need to disable this service:
reflector.service
ModemManager.service
systemd-resolved.service
systemd-networkd.socket
systemd-networkd.service

Just make sure that iwd and wpa-supplicant don't have access to the interface.

It is not normal that the build-in wifi card has not been enabled on boot.
Check dmesg to figure out why it failed.

It should look like this (I have a Realtek 8821ce PCIe card):
```
$ sudo dmesg
[    8.415437] cfg80211: Loading compiled-in X.509 certificates for regulatory database
...
[    9.166035] rtw_8821ce 0000:04:00.0: enabling device (0000 -> 0003)
[    9.167827] rtw_8821ce 0000:04:00.0: Firmware version 24.8.0, H2C version 12
...
[    9.238752] rtw_8821ce 0000:04:00.0 wlp4s0: renamed from wlan0
```

**Reply by v74863** [reply]
> Just make sure that iwd and wpa-supplicant don't have access to the interface

How exactly would you advise me to do that?

I found link
https://unix.stackexchange.com/questions/255313/detecting-which-application-is-using-which-network-interface

BTW: Have you heard the story below. It happened to my main laptop, too :(
https://h30434.www3.hp.com/t5/Notebook-Boot-and-Lockup/Bios-update-crash-problem-HP-ProBook-445-G7/td-p/9081020
https://borncity.com/win/2024/06/11/bios-update-01-17-00-bricks-hp-probooks-445-g7-and-455-g7/

**Comment by ZerBea** [comment]
I have not heard about the problems of HP notebooks, because I don't use them.
But I have absolutely no problems with my ASUS notebooks my MSI mainboard desktops and my Raspberry Pi Zero (not W/WH) systems.

I don't know the default configuration of the Arch ISO, but I guess it is completely different to my system configuration.
So I can't help you here, because (on my notebooks and desktops):
I use systemd-boot instead of grub
by default only tools that I really need are installed
by default only services that I really need are running
and everything is configured manually

Unwanted services can be disabled/stopped by systemctl.

BTW:
There is no "all-in-one solution suitable for every purpose"!
OpenWRT on the Raspberry is much smaller (and faster) than Arch on notebooks and desktops. The entire img.gz file (custom OpenWRT 23.05) of the Raspberry is only 58MB in size. It is exactly tailored to run hcxlabtool/hcxdumptool (nothing more), while the Arch Linux systems are tailored to do daily tasks (coding, analyzes which include Wireshark/tshark, hashcat and JtR tasks, home office and some multimedia stuff).

**Reply by v74863** [reply]
Let me ask more.
If you need to run hcxdumptool on an Arch Linux laptop, how do you configure that laptop? Please share the Author's configuration.
In particular, does it make sense to use the Realtime kernel?
https://wiki.archlinux.org/title/Kernel

> OpenWRT on the Raspberry is much smaller (and faster) than Arch on notebooks and desktops.

You said "faster". How do you measure? Please share the results of your measurements.

**Comment by ZerBea** [comment]
Please make up you own mind:
Raspbian Lite:
```
    Release date: July 4th 2024
    System: 32-bit
    Kernel version: 6.6
    Debian version: 12 (bookworm)
    Size: 500MB
```
https://downloads.raspberrypi.com/raspios_lite_armhf/images/raspios_lite_armhf-2024-07-04/2024-07-04-raspios-bookworm-armhf-lite.img.xz
A lot of services are enabled by default - they need to be disabled.

OpwnWRT:
$ ls -All
-rw-r--r-- 1 zerobeat users 61249184 10. Aug 21:07 20240810_bcm2708_config.img.gz
Only services which are mandatory are enabled by default.

Raspbian 500MB vs. OpenWRT 58MB

Take a stop watch and measure boot time OpenWRT vs Raspbian

For fun:
Take a stop watch and measure the time elapsed to install and configure OpenWRT.
Take a stop watch and measure the time elapsed to install and configure Raspbian.


I'm running hcxdumptool, hcxtools, hashcat and JtR on an ASUF TUF notebook. This is my configuration:
```
sudo dd bs=4M if=archlinux-2024.07.01-x86_64.iso of=/dev/sda status=progress oflag=sync


$ loadkeys de-latin1
================================================================================
$ lsblk
$ parted /dev/nvme0n1
(parted) mklabel gpt
(parted) mkpart "BOOT" fat32 1MiB 1026MiB
(parted) set 1 esp on
(parted) mkpart "SWAP" linux-swap 1026MiB 19408MiB (1026 + 16GB)
(parted) mkpart "ROOT" ext4 19408MiB 100%

$ mkfs.fat -F32 /dev/nvme0n1p1
$ mkswap /dev/nvme0n1p2
$ swapon /dev/nvme0n1p2
$ mkfs.ext4 /dev/nvme0n1p3

parted /dev/xxx
(parted) mklabel gpt
(parted) mkpart "HOME" 1MiB 100%
$ mkfs.ext4 /dev/xxx
================================================================================
$ mount /dev/nvme0n1p3 /mnt
$ mkdir -p /mnt/boot
$ mount /dev/nvme0n1p1 /mnt/boot

mkdir -p /mnt/home
mount /dev/xxx /mnt/home
================================================================================
$ pacstrap /mnt base base-devel linux linux-lts linux-firmware amd-ucode nano (if needed: efibootmgr)
$ genfstab -U /mnt > /mnt/etc/fstab
$ arch-chroot /mnt
================================================================================
$ ln -sf /usr/share/zoneinfo/Europe/Berlin /etc/localtime

$ hwclock --systohc

$ nano /etc/locale.gen
de_DE.UTF-8 UTF-8
de_DE ISO-8859-1
de_DE@euro ISO-8859-15

$ locale-gen

$ nano /etc/locale.conf
LANG=de_DE.UTF-8
$ echo "LANG=de_DE.UTF-8" > /etc/locale.conf

$ nano /etc/vconsole.conf
KEYMAP=de-latin1
FONT=eurlatgr
$ echo "KEYMAP=de-latin1" > /etc/vconsole.conf
$ echo "eurlatgr" >> /etc/vconsole.conf

$nano /etc/hostname
$ echo "archtop" > /etc/hostname
================================================================================
$ bootctl --path=/boot$esp install

nano /boot/loader/entries/arch.conf
title    Arch Linux
linux    /vmlinuz-linux
initrd   /amd-ucode.img
initrd   /initramfs-linux.img
options  root=UUID=..... rw lang=de init=/usr/lib/systemd/systemd locale=de_DE.UTF-8

nano /boot/loader/entries/arch-fallback.conf
title    Arch Linux Fallback
linux    /vmlinuz-linux
initrd   /initramfs-linux-fallback.img
options  root=UUID=..... rw lang=de init=/usr/lib/systemd/systemd locale=de_DE.UTF-8

nano /boot/loader/entries/arch-lts.conf
title    Arch Linux LTS
linux    /vmlinuz-linux-lts
initrd   /amd-ucode.img
initrd   /initramfs-linux-lts.img
options  root=UUID=..... rw lang=de init=/usr/lib/systemd/systemd locale=de_DE.UTF-8

nano /boot/loader/entries/arch-lts-fallback.conf
title    Arch Linux LTS Fallback
linux    /vmlinuz-linux-lts
initrd   /initramfs-linux-lts-fallback.img
options  root=UUID=..... rw lang=de init=/usr/lib/systemd/systemd locale=de_DE.UTF-8

nano /boot/loader/loader.conf 
default   arch.conf
timeout   4

bootctl update
================================================================================
$ useradd -m -g users -s /bin/bash ........
$ passwd ........

$ pacman -S sudo acpid ntp cups 
$ systemctl enable acpid.service
$ systemctl enable cups.service 
$ systemctl enable ntpd.service


$ nano /etc/ntp.conf
server de.pool.ntp.org
$ ntpd -gq
$ date
$ hwclock -w

$ export EDITOR=nano
$ export VISUAL=nano

$ visudo
#%sudo ALL=(ALL) ALL > %sudo ALL=(ALL) ALL
Defaults timestamp_timeout=15

$ groupadd sudo
$ gpasswd -a ........ sudo

$ pacman -S xf86-video-amdgpu nvidia nvidia-lts nvidia-settings opencl-nvidia cuda opencl-headers libxnvctrl xfce4 xfce4-goodies lightdm lightdm-gtk-greeter networkmanager xfce4-notifyd network-manager-applet 

systemctl enable lightdm.service
systemclt enable NetworkManager.service

$ sudo nano /etc/mkinitcpio.d/linux.preset
#ALL_microcode=(/boot/*-ucode.img)

passwd

exit
umount -R /mnt
reboot

localectl --no-convert set-keymap de-latin1
localectl --no-convert set-x11-keymap de pc105

sudo gpasswd -a ........ adbusers
sudo gpasswd -a ........ storage
sudo gpasswd -a ........ wireshark
sudo gpasswd -a ........ uucp
sudo gpasswd -a ........ scanner

git config --global core.editor "nano"

~/.bashrc
export EDITOR=nano
export VISUAL=nano

pacman -S gvfs gvfs-smb gvfs-mtp man usbutils make cmake gcc man man-pages-de fwupd htop bcc bcc-tools bpf udisks2
pacman -S hicolor-icon-theme ttf-freefont ttf-liberation ttf-dejavu pacman-contrib linux-headers linux-firmware dialog bc cpio yasm nasm cmake valgrind reflector strace afl squashfs-tools radare2 gst-plugins-bad gst-plugins-ugly dvdbackup smartmontools inetutils
pacman -S libmtp smbclient libpcap socat wireless-regdb iw iproute2 nmap ncrack tcpdump dsniff sqlmap ngrep hydra netsniff-ng wireshark-qt fping bluez traceroute iperf handbrake avidemux-qt android-tools android-udev clamav rkhunter asunder
pacman -S hdparm ddrescue testdisk lsof lsdvd lshw vdpauinfo alsa-utils pipewire-alsa pipewire-pulse pavucontrol dmidecode subversion git rsync mlocate fdupes dos2unix html2text binwalk ghex geany geany-plugins galculator openssh sshfs mc sqlite3 sqlitebrowser mkvtoolnix-cli
pacman -S gtk3-print-backends falkon firefox firefox-i18n-de firefox-noscript firefox-ublock-origin  w3m lynx wget pidgin filezilla transmission-gtk yt-dlp rtmpdump thunderbird thunderbird-i18n-de hunspell-de hunspell-fr hunspell-en_US
pacman -S viking jdk-openjdk gpsbabel gparted gpart btrfs-progs dosfstools exfatprogs ntfs-3g mtools p7zip unrar unzip unarj unace lzop xarchiver hplip xsane geeqie imagemagick jhead libopenraw gimp gphoto2 perl-image-exiftool perl-archive-zip qtqr qrencode zbar tesseract-data-deu
pacman -S libgepub libdvdcss mediainfo cdrtools vorbis-tools lame audacious mpv audacity twolame smplayer ffmpeg xclip calibre bcprov pdftk qpdf mupdf xpdf poppler-data libgxps libreoffice-fresh libreoffice-fresh-de
```

20-amd-ryzen.conf (to make sure the GUI use it instead of the build in NVIDIA GPU
)```
Section "ServerLayout"
    Identifier     "Layout0"
    Screen 0       "amd" 0 0
EndSection

Section "Device"
    Identifier    "amd"
    Driver        "amdgpu"
    Screen        0
EndSection

Section "Screen"
    Identifier     "amd"
    Device         "amddev"
EndSection
```

**Reply by v74863** [reply]
> Raspbian 500MB vs OpenWRT 58MB

Ok, impressive, accepted.

> boot time Raspbian vs OpenWRT

Ok, accepted.

But when both systems are fully booted, are there any noticeable differences?

I think it is desirable to have benchmark specifically designed for hcxdumptool.
 inspired by https://www.cpubenchmark.net

BTW: Have you ever tried to use hcxdumptool with Arch linux Realtime kernel?
https://wiki.archlinux.org/title/Realtime_kernel_patchset

**Comment by ZerBea** [comment]
Basically OpenWRT is a Linux distribution and it is running a Linux kernel like all other Linux distributions, too. So far the basics.
In a Linux distribution there are thousands of packages which you don't need on a router (GUIs like XFCE, LXQT, KDE, GNOME, web browsers like firefox and a lot more). Unfortunately this packages are installed, configured and their services are enabled.
While Linux distributions expects users to interact with the software using a screen and keyboard or mouse, these capabilities are not needed on router. You don't need all those input and display drivers and services that take a lot of CPU cycles.

For sure, you can install a Linux distribution and remove everything you don't need, configure it to run headless and disable all service you don't want. That include to reconfigure and to recompile the Linux kernel so that it is tailored to run as router.

Or you get the OpenWRT toolchain and compile a system that is tailored exactly to your hardware (e.g. Raspberry Pi Zero & WiFi adapter). The entire system is tailored to be a router. Everything that take additional CPU cycles has been compiled out.

Please take a look at my OpenWRT .config file:
[config.zip](https://github.com/user-attachments/files/16648333/config.zip)
You'll notice hundreds settings that are disabled, because they are useless for my task, but take a lot of CPU cycles if compiled in.
I guess I've more settings disabled than enabled - and there is still room for some improvements.
Please take a look at line 1698:
hcxlabtool is compiled directly into the image:
```
#
# Custom
#
CONFIG_PACKAGE_hcxlabtool=y
# end of Custom
```

To save power, config.txt has been modified, too:
```
# OpenWrt config
include distroconfig.txt

[all]
# Place your custom settings here.
gpu_mem=16
disable_splash=1
dtparam=audio=off
dtoverlay=hdmi_blanking=1
[pi0]
dtparam=act_led_trigger=none
dtparam=act_led_activelow=off
[pi0w]
dtoverlay=disable-wifi
dtoverlay=disable-bt
[pi1]
dtparam=pwr_led_trigger=none
dtparam=pwr_led_activelow=off
```
On a standard distribution most of this useless stuff is enabled by default and you have to wrestle with this configurations if something is not working as you expect.


I've never used the Arch Realtime Kernel.
Before I shut down my systems, I do:
```
$ sudo pacman -Syu
$ sudo pacman -Scc
$ pacman -Qtdq -> to identify packages that can be removed
$ sudo journalctl --vacuum-time=10d
```

**Reply by v74863** [reply]
Thank you for detailed reply. I appreciate it very much.
However, it is still unclear for me are there any noticeable differences between different systems in the success rate of the attack once the systems have fully booted up and begun to execute the attack cycle?

Or the image size and boot time are the only noticeable differences?

Let me clarify what I mean.
Let's take two systems.
1st - Raspberry Pi with OpenWRT,
2nd - your ASUF TUF notebook with Arch Linux.

For the attack to be successful, hcxdumptool must respond to a client request faster than a legitimate access point. A legitimate access point respond also, but a little later.

Thus, there is a time delta. Have you measured this delta on different systems?
Which system is the winner? In other words, which system responds faster?

To an outside observer, our systems are black boxes. The differences in the operation of these systems for this observer will probably consist only in different time delays between requests and responses.

**Comment by ZerBea** [comment]
You can't compare the 2 systems because the RPI is a single core and the TUF is a quad core running 8 threads.
If hcxdumptool is running on one of the cores of the TUF there are still 3 cores remaining to handle other tasks.
That is not the case on the RPI.
Basically the success rate (attacked targets in e.g. 10 seconds) is the same on both systems.
hcxdumptool is highly optimized to run on every system (this time an 8 core ryzen) running a GUI and hashcat in parallel:
```
   6401 zerbea   20   0   16,8g   1,9g  86564 S   7,3   3,2   3:16.11 hashcat                                                              
    668 root      20   0 1345952 193728 120916 S   0,7   0,3   1:30.16 Xorg                                                                 
   8720 root       20   0    5008   2908   2396 S   0,3   0,0   0:00.12 hcxdumptool                                                          
   9104 zerbea  20   0   10532   7512   5336 R   0,3   0,0   0:00.03 top                                                                  
```
only 23 packets are dropped by the kernel:
```
9328 Packet(s) captured by kernel
23 Packet(s) dropped by kernel
1 SHB written to pcapng dumpfile
1 IDB written to pcapng dumpfile
1 ECB written to pcapng dumpfile
191 EPB written to pcapng dumpfile
```


Running on a RPI Zero at the same time with a GUI and firefox, I don't think that hcxdumptool is able to update all its internal ring buffers in case of high traffic - but I have not measured this.

**Comment by ZerBea** [comment]
Added some time measurement e.g entire function to process an incoming BEACON frame:
`process80211beacon() process took 0.000027 seconds to execute`
that include sorting the internal ring buffer on every new beacon

Depending on the traffic, hcxdumptool has to do this sereral times a second:
```
$ tshark -r xx.pcapng -Y "wlan.fc.type_subtype == 0x08" -T fields -e frame.time | cut -c 20-
...
21.983724182 CEST
22.086103456 CEST
22.188513351 CEST
22.211085657 CEST
22.211445649 CEST
22.290681314 CEST
22.313452892 CEST
22.393099362 CEST
22.518681743 CEST
22.598097111 CEST
22.621441829 CEST
22.700657176 CEST
22.723302950 CEST
22.802937598 CEST
22.825431618 CEST
22.826106227 CEST
22.905343714 CEST
23.007724242 CEST
...
```
In case of processing BEACONs, we are in good shape.
```
22.211085657 CEST 22.211445649 CEST == 0,000359992 elapsed time between 2 BEACONs
0.000027 seconds to process a BEACON
```
There is enough time to process the BEACON and to start an attack (request PMKID).
But I' sure you can imagine what happens on a single core machine which is running several other tasks at the same time.

**Comment by ZerBea** [comment]
If you're interested in how I do the measurement. This is the function to measure the elapsed time:
```
static clock_t t; 
...
	if(macfrx->subtype == IEEE80211_STYPE_BEACON)
		{
		t = clock(); 
		process80211beacon();
		t = clock() - t; 
		double time_taken = ((double)t)/CLOCKS_PER_SEC; // in seconds 
		printf("process80211beacon() took %f seconds to execute \n", time_taken); 
		}
```
